environments/servlet/core/src/main/java/org/jboss/weld/environment/servlet/deployment/WebAppBeanArchiveScanner.java

```diff
@@ -100,18 +92,34 @@ public WebAppBeanArchiveScanner(ResourceLoader resourceLoader, Bootstrap bootstr
             if (beansXmlUrl != null) {
                 BeansXml beansXml = parseBeansXml(beansXmlUrl);
                 if (accept(beansXml)) {
-                    File webInfClasses = Servlets.getRealFile(servletContext, WEB_INF_CLASSES);
+                    webInfClasses = Servlets.getRealFile(servletContext, WEB_INF_CLASSES);
                     if (webInfClasses != null) {
-                        results.add(new ScanResult(beansXml, webInfClasses.getPath()).extractBeanArchiveId(contextPath, WEB_INF));
+                        webInfBeansXML = new ScanResult(beansXml, webInfClasses.getPath()).extractBeanArchiveId(contextPath, WEB_INF);
                     } else {
                         // The WAR is not extracted to the file system - make use of ServletContext.getResourcePaths()
-                        results.add(new ScanResult(beansXml, WEB_INF_CLASSES));
+                        webInfBeansXML = new ScanResult(beansXml, WEB_INF_CLASSES);
                     }
                 }
             }
         } catch (MalformedURLException e) {
             throw WeldServletLogger.LOG.errorLoadingResources(e);
         }
+
+        // All previous results for WEB-INF/classes must be ignored
+        for (Iterator<ScanResult> iterator = results.iterator(); iterator.hasNext();) {
+            ScanResult result = iterator.next();
+            String path = result.getBeanArchiveRef().toString();
+            if (new File(path).equals(webInfClasses)) {
```

Comment:
`webInfClasses` is `null` for WARs that are not extracted to the file system. I believe that we should also keep the original conditions, i.e. :
```java
if (path.contains(WEB_INF_CLASSES_FILE_PATH) || path.contains(WEB_INF_CLASSES) || new File(path).equals(webInfClasses)) {
 ...
}
```

Reply:
I wonder whether the DefaultBeanArchiveScanner would find a beans.xml in a non-extracted WAR.
Keeping the original conditions could lead to unwanted side-effects, in the unlikely case that someone calls his build directory WEB-INF. (In fact, I did that as an attempt of a quick and dirty workaround before solving it properly). We should find a condition that only excludes the _previously found_ beans.xml, as stated by the comment above the loop.

Reply:
However, I guess for now, any path containing WEB-INF/classes not actually being the server´s WEB-INF/classes directory can be considered a rare edge case.